osquery/events/windows/ntfs_event_publisher.cpp

```diff
@@ -0,0 +1,615 @@
+      ::GetFinalPathNameByHandle(handle,
+                                 &buffer[0],
+                                 static_cast<DWORD>(buffer.size()),
+                                 FILE_NAME_NORMALIZED | VOLUME_NAME_DOS));
+
+  auto error_code = ::GetLastError();
+  ::CloseHandle(handle);
+
+  if (bytes_returned == 0U || bytes_returned >= buffer.size()) {
+    std::stringstream message;
+    message << "Failed to acquire the path for the file in volume "
+            << drive_letter << ":\\. Error: ";
+
+    std::string description;
+    if (!getWindowsErrorDescription(description, error_code)) {
+      description = "Unknown error";
+    }
+
+    message << description;
+    return Status::failure(message.str());
+  }
+
+  // Paths follow this form: \\?\C:\\path\\to\\folder; skip the prefix
+  path = buffer.c_str() + 4;
+  buffer.clear();
+
+  return Status::success();
+}
+
+Status NTFSEventPublisher::getPathFromParentFRN(
+    std::string& path,
+    ParentFRNCache& parent_frn_cache,
+    char drive_letter,
+    const std::string& basename,
+    const USNFileReferenceNumber& ref) {
+  auto& it = parent_frn_cache.find(ref);
+
+  if (it == parent_frn_cache.end()) {
+    Status status = getPathFromReferenceNumber(path, drive_letter, ref);
+
+    if (!status.ok()) {
+      return status;
+    }
+
+    parent_frn_cache[ref] = path;
+  } else {
+    path = it->second;
+  }
+
+  path.append("\\" + basename);
+
+  return Status::success();
+}
+
+Status NTFSEventPublisher::getVolumeData(VolumeData& volume,
+                                         char drive_letter) {
+  UpgradeLock lock(d_->volume_data_map_mutex);
+
+  {
+    auto it = d_->volume_data_map.find(drive_letter);
+    if (it != d_->volume_data_map.end()) {
+      volume = it->second;
+      return Status::success();
+    }
+  }
+
+  WriteUpgradeLock wlock(lock);
+
+  // Get a handle to the volume
+  auto volume_path = std::string("\\\\.\\") + drive_letter + ":";
+
+  VolumeData volume_data = {};
+  volume_data.volume_handle =
+      ::CreateFile(volume_path.c_str(),
+                   GENERIC_READ,
+                   FILE_SHARE_READ | FILE_SHARE_WRITE | FILE_SHARE_DELETE,
+                   nullptr,
+                   OPEN_EXISTING,
+                   FILE_FLAG_BACKUP_SEMANTICS,
+                   nullptr);
+
+  if (volume_data.volume_handle == INVALID_HANDLE_VALUE) {
+    std::stringstream message;
+    message << "Failed to open the following drive: " << volume_path
+            << " due to the following error: ";
+
+    std::string description;
+    if (!getWindowsErrorDescription(description, ::GetLastError())) {
+      description = "Unknown error";
+    }
+
+    message << description;
+    return Status::failure(message.str());
+  }
+
+  // Get the root folder reference number
+  std::string root_folder_path;
+  root_folder_path.push_back(drive_letter);
+  root_folder_path.append(":\\");
+
+  volume_data.root_folder_handle =
```

Comment:
I mentioned `scope_guard`s above, they might be pretty handy for some of these `HANDLE`s being stored by the `volume_data`, but then again the guard `CloseHandle`s as soon as it goes out of scope, so I'm not totally sure if it would work or not. Have we checked this table impl generically to ensure there aren't any handle leaks?